dali/operators/sequence/element_extract.h

```diff
@@ -18,50 +18,87 @@
 
 #include <vector>
 #include "dali/core/common.h"
+#include "dali/core/format.h"
+#include "dali/core/tensor_layout.h"
 #include "dali/core/tensor_shape.h"
+#include "dali/kernels/common/scatter_gather.h"
 #include "dali/pipeline/operator/operator.h"
 
 namespace dali {
 
 namespace detail {
-  static void CheckInputShape(const TensorShape<>& tensor_shape,
+  static void CheckInputShape(const span<const int64_t>& tensor_shape,
                               const std::vector<int>& element_map) {
-    DALI_ENFORCE(tensor_shape.size() > 0);
+    assert(tensor_shape.size() > 0);
     auto N_input = tensor_shape[0];
 
-    int N_output = element_map.size();
-    DALI_ENFORCE(N_input >= N_output,
-        "Requested more elements than available");
-
     for (auto elem : element_map)
-        DALI_ENFORCE(elem < N_input,
-            "index " + std::to_string(elem) + " out of bounds");
+      DALI_ENFORCE(elem < N_input, "index " + std::to_string(elem) + " out of bounds");
+  }
+
+  static TensorListShape<> GetOutputShape(const TensorListShape<> &input_shape,
+                                          const std::vector<int> &element_map,
+                                          const TensorLayout& input_layout) {
+
+    if (!input_layout.empty()) {
+      DALI_ENFORCE(
+          VideoLayoutInfo::IsSequence(input_layout),
+          make_string("Input layout must describe a sequence - it must start with 'F', got '",
+                      input_layout, "' instead."));
+    }
+
+    DALI_ENFORCE(input_shape.sample_dim() > 1,
+                 "Input must have at least two dimenstions - outermost for sequence and at least "
+                 "one for data elements.");
+
+    for (int i = 0; i < input_shape.num_samples(); ++i) {
+      auto shape = input_shape.tensor_shape_span(i);
+      CheckInputShape(shape, element_map);
+    }
+    return input_shape.last(input_shape.sample_dim() - 1);
   }
+
 }  // namespace detail
 
 template <typename Backend>
 class ElementExtract : public Operator<Backend> {
  public:
   inline explicit ElementExtract(const OpSpec &spec)
-    : Operator<Backend>(spec) {
+      : Operator<Backend>(spec), scatter_gather_(kMaxSizePerBlock) {
     element_map_ = spec.GetRepeatedArgument<int>("element_map");
 
-    DALI_ENFORCE(!element_map_.empty(),
-        "No 'element_map' indexes provided");
+    DALI_ENFORCE(!element_map_.empty(), "No 'element_map' indexes provided");
+
+    for (auto elem : element_map_) {
+      DALI_ENFORCE(elem >= 0, "index " + std::to_string(elem) + " out of bounds.");
```

Comment:
I would say that the value is negative because you don't check the upper bound here,

Reply:
Done